lib/spack/spack/spec.py

```diff
@@ -4557,11 +4559,20 @@ def parse_compiler(self, text):
         self.setup(text)
         return self.compiler()
 
+    @staticmethod
+    def _lookup_local_or_remote_hash(find_hash):
+        # type: (str) -> List[Spec]
+        return (spack.store.db.get_by_hash(find_hash) or
+                # If the hash doesn't exist locally, check the remote.
+                spack.binary_distribution.binary_index.find_prefix_hash(find_hash))
```

Comment:
@becker33, I think this is what you were saying last week that we *cannot* do, i.e. check the database and then the binary mirrors sequentially, when we look for matches for a hash prefix.  If I understood your argument correctly, you were saying if we search those places for a hash prefix sequentially, we could end up in a situation where we mistakenly return a spec from the install database (because we found there exactly one spec with the queried hash prefix), when we should have raised an ambiguous hash exception (because the binary mirrors or active env contain specs with the same hash prefix, possibly even the spec the user *actually meant* when they abbreviated the full hash with a prefix).

@cosmicexplorer I came here because I, too, want the spec parser to look somewhere other than just the install database for hashes.  And I made the same sequential searching mistake in my PR [here](https://github.com/spack/spack/pull/22657/files#diff-9e28cdd6f9cd66e61101de6e9d22b1c33561959e4305c63147bbfbf7b22da936R4566).  Where you are trying to get spack to include all binary specs in its hash search, my goal is to have it include all specs from the active concrete environment.  So I'm looking to create some kind of dictionary to map hashes to specs which will be populated from all the locations where hashes can exist: db, binary mirrors, active environment (if concrete).